### PR TITLE
Implement Fill layers for Metal

### PR DESCRIPTION
**Pull request by alanchenboy**
Implement Fill layers for Metal

Simulator Screen Shot - iPhone 14 Plus - 2023-08-11 at 17 32 17

Cause the fill bucket doesn't support triangulating the outline into triangles, it can only be drawn as lines, and in Metal there is no way to draw thick lines with line primitive. We need to triangulate the outline into triangles in the future and achieve the same effect as OpenGL.

Cause textures are not finished before this PR, the fill pattern is not verified.

### CODE REVIEW

**Review by sjg-wdw — COMMENTED** [review]
Looks pretty good!
One question I have in general is the various Metal #ifdefs 
I realize we're only doing permutations this way on the Metal side, but can we make that a bit more generic?  Perhaps check for a permutation state on the rendering context and add the UBOs accordingly.

If this is how we're doing it for now, let's merge it in and then revisit the issue later.

File: src/mbgl/renderer/layers/render_fill_layer.cpp [hunk 1]
```diff
@@ -366,7 +377,7 @@ void RenderFillLayer::update(gfx::ShaderRegistry& shaders,
     stats.drawablesRemoved += tileLayerGroup->removeDrawablesIf([&](gfx::Drawable& drawable) {
         // If the render pass has changed or the tile has dropped out of the cover set, remove it.
         const auto& tileID = drawable.getTileID();
-        if (drawable.getRenderPass() != passes || (tileID && !hasRenderTile(*tileID))) {
+        if ((drawable.getRenderPass() | passes) == RenderPass::None || (tileID && !hasRenderTile(*tileID))) {
```
[the remark this answers is not in the document]
Comment (alanchenboy) [reply src/mbgl/renderer/layers/render_fill_layer.cpp:380]: Thanks for review. this is a mistake.

[the remark this answers is not in the document]
Comment (alanchenboy) [reply src/mbgl/renderer/layers/render_fill_layer.cpp]: this line fixed

**Comment by alanchenboy** [comment]
How can I remove all these #ifdefs, any example? or use a flag to run it on runtime?

**Comment by alexcristici** [comment]
> How can I remove all these #ifdefs, any example? or use a flag to run it on runtime?

You should use `permutationUBO` concept and you can remove all the #ifdefs.

**Comment by sjg-wdw** [comment]
For the fill outlines let's just use the Metal line implementation for the moment.  That won't have the right width, but we can circle back and address it later.
The issue is that wide line reuse is something of a bucket reuse problem.  Something we'd like to address more generally.

**Comment by alanchenboy** [comment]
> > How can I remove all these #ifdefs, any example? or use a flag to run it on runtime?
> 
> You should use `permutationUBO` concept and you can remove all the #ifdefs.

Hi Alex, I already use `FillPermutationUBO`, `FillOutlinePermutationUBO`, etc. There are no `#ifdefs` in shaders, can I directly remove #ifdefs in cpp code like `#if MLN_RENDER_BACKEND_METAL`

**Comment by alexcristici** [comment]
> > > How can I remove all these #ifdefs, any example? or use a flag to run it on runtime?
> > 
> > 
> > You should use `permutationUBO` concept and you can remove all the #ifdefs.
> 
> Hi Alex, I already use `FillPermutationUBO`, `FillOutlinePermutationUBO`, etc. There are no `#ifdefs` in shaders, can I directly remove #ifdefs in cpp code like `#if MLN_RENDER_BACKEND_METAL`

For now we should keep the #ifdefs in the cpp code, in order to switch between Metal and GL. Maybe we will refactor it later after we finish implementing all the layers.

File: include/mbgl/shaders/mtl/fill.hpp [hunk 2]
```diff
@@ -0,0 +1,472 @@
...
+        {2, gfx::AttributeDataType::Float2, 1, "a_opacity"},
+    };
+    static constexpr UniformBlockInfo uniforms[] = {
+        MLN_MTL_UNIFORM_BLOCK(3, true, false, FillDrawableUBO),
+        MLN_MTL_UNIFORM_BLOCK(4, true, true, FillEvaluatedPropsUBO),
+        MLN_MTL_UNIFORM_BLOCK(5, true, false, FillInterpolateUBO),
+        MLN_MTL_UNIFORM_BLOCK(6, true, true, FillPermutationUBO),
+        MLN_MTL_UNIFORM_BLOCK(7, true, false, ExpressionInputsUBO),
+    };
+
```
Comment (alexcristici) [inline include/mbgl/shaders/mtl/fill.hpp:29]: The `textures` are missing and generates compiling error.

**Review by stefankarschti — APPROVED** [review]

**Review by TimSylvester — APPROVED** [review]

**Comment by TimSylvester** [comment]
Too late now, but I didn't notice that you reworked all the UBOs.  Unfortunately, that means all the OpenGL shaders need to be reworked now.

**Comment by alanchenboy** [comment]
@TimSylvester  Let me try open gl compiling.

**Pull request merged**